File: src/python/nispor/macsec.py
# SPDX-License-Identifier: Apache-2.0

from .base_iface import NisporBaseIface


class NisporMacsec(NisporBaseIface):
    def __init__(self, info):
        super().__init__(info)
        self._macsec_info = self._info.get("macsec", {})

    @property
    def sci(self):
        return self._macsec_info["sci"]

    @property
    def port(self):
        return self._macsec_info["port"]

    @property
    def icv_len(self):
        return self._macsec_info["icv_len"]

    @property
    def cipher(self):
        return self._macsec_info["cipher"]

    @property
    def window(self):
        return self._macsec_info["window"]

    @property
    def encoding_sa(self):
        return self._macsec_info["encoding_sa"]

    @property
    def encrypt(self):
        return self._macsec_info["encrypt"]

    @property
    def protect(self):
        return self._macsec_info["protect"]

    @property
    def send_sci(self):
        return self._macsec_info["send_sci"]

    @property
    def end_station(self):
        return self._macsec_info["end_station"]

    @property
    def scb(self):
        return self._macsec_info["scb"]

    @property
    def replay_protect(self):
        return self._macsec_info["replay_protect"]

    @property
    def validate(self):
        return self._macsec_info["validate"]

    @property
    def offload(self):
        return self._macsec_info["offload"]

    @property
    def base_iface(self):
        return self._macsec_info["base_iface"]

File: src/lib/integ_tests/macsec.rs
// SPDX-License-Identifier: Apache-2.0

use std::panic;

use pretty_assertions::assert_eq;

use crate::NetState;

use super::utils::assert_value_match;

const IFACE_NAME: &str = "macsec0";

// SCI has been excluded from the state as it is non-deterministic
const EXPECTED_MACSEC_INFO: &str = r#"---
name: macsec0
iface_type: mac_sec
macsec:
  port: 0
  icv_len: 16
  cipher: gcm-aes128
  window: 0
  encoding_sa: 0
  encrypt: true
  protect: true
  send_sci: true
  end_station: false
  scb: false
  replay_protect: false
  validate: strict
  base_iface: eth1"#;

#[test]
fn test_get_macsec_iface_yaml() {
    with_macsec_iface(|| {
        let state = NetState::retrieve().unwrap();
        let iface = &state.ifaces[IFACE_NAME];

        assert_eq!(iface.iface_type, crate::IfaceType::MacSec);

        assert_value_match(EXPECTED_MACSEC_INFO, iface);
    });
}

fn with_macsec_iface<T>(test: T)
where
    T: FnOnce() + panic::UnwindSafe,
{
    super::utils::set_network_environment("macsec");

    let result = panic::catch_unwind(|| {
        test();
    });

    super::utils::clear_network_environment();
    assert!(result.is_ok())
}
